### PR TITLE
Add Visualization chat commands

### LINKED ISSUE #13252: Debug menu for spectators ?
Watching 100 % AI games is very helpful to find bugs.
As a spectator/admin/host, if you enable the debug menu in the lobby then you don't have it ingame.
Could it be possible to add a special debug menu for spectators (with Vizualizations only) ?

debug_menu_enabled

no_debug_menu

### PR DESCRIPTION
**Pull request by rob-v**
New visualization chat commands - showcombatgeometry, showdebuggeometry, showdepthpreview, showactortags - available also for spectators.

Show Unit Paths (DevPathDebug chat command) is in Visualization section, but is also player command, so not available for spectators.
Were already available: terrainoverlay, debugcustomterrain.

To enable these 4 options in Debug window, World needs VisualizationMode trait. To enable also chat commands, World needs also VisualizationCommands trait.

Closes #13252

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs [hunk 1]
```diff
@@ -25,12 +25,16 @@ public class CombatDebugOverlayInfo : ITraitInfo
 		public object Create(ActorInitializer init) { return new CombatDebugOverlay(init.Self); }
 	}
 
-	public class CombatDebugOverlay : IRenderAboveWorld, INotifyDamage, INotifyCreated
+	public class CombatDebugOverlay : IRenderAboveWorld, INotifyDamage, INotifyCreated, IWorldLoaded, IChatCommand
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs]: Every actor in the world has its own copy of this trait, so I don't expect this will work properly.

File: OpenRA.Mods.Common/Traits/Render/RenderDebugState.cs [hunk 2]
```diff
@@ -27,14 +28,18 @@ class RenderDebugStateInfo : ITraitInfo
 		public object Create(ActorInitializer init) { return new RenderDebugState(init.Self, this); }
 	}
 
-	class RenderDebugState : INotifyAddedToWorld, INotifyOwnerChanged, IRenderAboveShroudWhenSelected
+	class RenderDebugState : INotifyAddedToWorld, INotifyOwnerChanged, IRenderAboveShroudWhenSelected, IChatCommand
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderDebugState.cs]: Likewise for this one.

**Comment by rob-v** [comment]
Updated - new PR description and implementation.

**Review by abcdefg30 — COMMENTED** [review]
You're removing fields from `DeveloperModeInfo`, but aren't re-adding them to `VisualizationModeInfo`. Is that intentional/wanted?

**Comment by rob-v** [comment]
It is intentional, not sure if wanted. I think (almost) always either you want those visualization commands or not. IMO it isn't needed/wanted to disable some of the visualization commands, you simply don't use them.

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by rob-v** [comment]
Rebased.

File: OpenRA.Game/Traits/World/VisualizationMode.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * information, see COPYING.
+ */
+#endregion
+
+namespace OpenRA.Traits
+{
+	[Desc("Enables visualization commands. Attach this to the world actor.")]
+	public class VisualizationModeInfo : TraitInfo<VisualizationMode> { }
+
+	public class VisualizationMode
```
Comment (pchote) [inline OpenRA.Game/Traits/World/VisualizationMode.cs]: Could we call these `DebugVisualizations(Info)`, and then `CombatGeometry`/`RenderGeometry`/`DepthBuffer`/`ActorTags`?  IMO these are a lot clearer and make the most of the context of now having our own dedicated container trait.

File: OpenRA.Mods.Common/Commands/VisualizationCommands.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+using System;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Commands
+{
+	[Desc("Enables visualization commands via the chatbox. Attach this to the world actor.")]
+	public class VisualizationCommandsInfo : TraitInfo<VisualizationCommands> { }
+
+	public class VisualizationCommands : IChatCommand, IWorldLoaded
```
Comment (pchote) [inline OpenRA.Mods.Common/Commands/VisualizationCommands.cs]: and then `DebugVisualizationCommands(Info)` here for consistency.

**Review by pchote — APPROVED** [review]
This works well.  Just a couple of naming suggestions that I think would make this a bit tidier.

One really great side effect of this is that the core engine no longer depends on the `DeveloperMode` trait.  Would you mind adding a commit that moves this to Mods.Common?

**Comment by rob-v** [comment]
Updated.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 5]
```diff
@@ -816,22 +817,14 @@
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
   <Target Name="AfterBuild">
-    <MakeDir Directories="$(SolutionDir)mods/common/"/>
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Looks like some unwanted VS reformatting slipped in. Can you please unstage these?

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 6]
```diff
@@ -816,22 +818,14 @@
...
-    <Copy
-      SourceFiles="$(TargetPath)"
-      DestinationFolder="$(SolutionDir)mods/common/"/>
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Those changes don't look related to this PR. Do we want to revert this? (I don't mind it really.)

Comment (rob-v) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Yes, I am reverting those VS formatting changes.

**Comment by rob-v** [comment]
Reverted VS formatting.

File: mods/cnc/rules/world.yaml [hunk 7]
```diff
@@ -6,6 +6,7 @@
 	MusicPlaylist:
 		VictoryMusic: win1
 		DefeatMusic: nod_map1
+	DebugVisualizations:
```
Comment (obrakmann) [inline mods/cnc/rules/world.yaml]: It would probably make sense to add this via an upgrade rule. What to you think?

**Review by obrakmann — APPROVED** [review]
Looks good otherwise, 👍

File: mods/cnc/rules/world.yaml (hunk 7 above)
Comment (pchote) [inline mods/cnc/rules/world.yaml]: Probably more trouble than its worth to stop it from adding these to maps that override World.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ef8502667775b878c09df5f54c4a0bb58af49391)
